File: src/main/java/com/github/anba/es6draft/runtime/types/Constructor.java
/**
 * Copyright (c) 2012-2015 André Bargull
 * Alle Rechte vorbehalten / All Rights Reserved.  Use is subject to license terms.
 *
 * <https://github.com/anba/es6draft>
 */
package com.github.anba.es6draft.runtime.types;

import com.github.anba.es6draft.runtime.ExecutionContext;

/**
 * <h1>6 ECMAScript Data Types and Values</h1><br>
 * <h2>6.1 ECMAScript Language Types</h2><br>
 * <h3>6.1.7 The Object Type</h3>
 * <ul>
 * <li>6.1.7.2 Object Internal Methods and Internal Slots
 * </ul>
 * <p>
 * Internal Method: [[Construct]]
 */
public interface Constructor extends ScriptObject, Callable {
    /**
     * [[Construct]]
     * 
     * @param callerContext
     *            the caller's execution context
     * @param newTarget
     *            the constructor to which the {@code new} operator was initially applied
     * @param args
     *            the constructor function arguments
     * @return the new script object
     */
    ScriptObject construct(ExecutionContext callerContext, Constructor newTarget, Object... args);

    /**
     * [[Construct]] in tail-call position
     * 
     * @param callerContext
     *            the caller's execution context
     * @param newTarget
     *            the constructor to which the {@code new} operator was initially applied
     * @param args
     *            the constructor function arguments
     * @return the new script object
     * @throws Throwable
     *             any error thrown by the underlying method implementation
     */
    Object tailConstruct(ExecutionContext callerContext, Constructor newTarget, Object... args)
            throws Throwable;
}

File: src/test/scripts/suite/objects/Set/constructor.js
/*
 * Copyright (c) André Bargull
 * Alle Rechte vorbehalten / All Rights Reserved.  Use is subject to license terms.
 *
 * <https://github.com/anba/es6draft>
 */
const {
  assertSame,
  assertThrows,
} = Assert;

// 23.2.1  The Set Constructor

// undefined and null are ignored
new Set(void 0);
new Set(null);

// TypeError for non-iterable
for (let nonIterable of [0, 1, 0.5, 0 / 0, true, false, Symbol.iterator, Symbol()]) {
  assertThrows(TypeError, () => new Set(nonIterable));
}
for (let nonIterable of [{}, {a: 0}, /(?:)/, new Date, () => {}]) {
  assertThrows(TypeError, () => new Set(nonIterable));
}

// No TypeError for iterable
for (let iterable of [[], [0, 1], function*(){}(), "", "abc"]) {
  new Set(iterable);
}
